### PR TITLE
Finodev

### PR DESCRIPTION
**Pull request by ashiq24**

### CODE REVIEW

File: neuralop/models/uno.py [hunk 1]
```diff
@@ -153,19 +153,25 @@ def __init__(self,
...
+            self.output_scaling_factor = np.multiply(self.output_scaling_factor, k)
+        self.output_scaling_factor = self.output_scaling_factor.tolist()
+        if len(self.output_scaling_factor) == 1:
+            self.output_scaling_factor = self.output_scaling_factor[0]
+
         if isinstance(self.output_scaling_factor, (float, int)):
             self.output_scaling_factor = [self.output_scaling_factor]*self.n_dim
         
-
+        print("calculated out factor", self.output_scaling_factor)
```
Comment (JeanKossaifi) [inline neuralop/models/uno.py:174]: Let's add an attribute `verbose` (pass it in init) and only print `if self.verbose:`

File: neuralop/models/uno.py [hunk 2]
```diff
@@ -195,7 +201,7 @@ def __init__(self,
                                             use_mlp=use_mlp, 
                                             mlp_dropout=mlp_dropout, 
                                             mlp_expansion=mlp_expansion,
-                                            output_scaling_factor = self.uno_scalings[i],
+                                            output_scaling_factor = [self.uno_scalings[i]],
```
Comment (JeanKossaifi) [inline neuralop/models/uno.py:204]: Why is it a list?

File: neuralop/models/uno.py [hunk 3]
```diff
@@ -186,7 +192,7 @@ def __init__(self,
 
             if i in self.horizontal_skips_map.keys():
                 prev_out = prev_out + self.uno_out_channels[self.horizontal_skips_map[i]]
-            print([self.uno_scalings[i]])
+            print(self.uno_scalings[i])
```
Comment (JeanKossaifi) [inline neuralop/models/uno.py:195]: I don't think this print is necessary

File: neuralop/models/uno.py [hunk 4]
```diff
@@ -225,6 +231,7 @@ def forward(self, x):
 
         if self.domain_padding is not None:
             x = self.domain_padding.pad(x)
+        print(x.shape)
```
Comment (JeanKossaifi) [inline neuralop/models/uno.py:234]: Let's remove this and the next print too

File: neuralop/models/uno.py [hunk 5]
```diff
@@ -153,19 +153,25 @@ def __init__(self,
...
         self.operator_block = operator_block
         self.integral_operator = integral_operator
         
         if self.horizontal_skips_map is None:
             self.horizontal_skips_map = {}
             for i in range(n_layers//2,0,):
                 self.horizontal_skips_map[n_layers - i -1] = i
-                
+
+        self.output_scaling_factor = np.ones_like(self.uno_scalings[0])
```
Comment (JeanKossaifi) [inline neuralop/models/uno.py:164]: Let's explicitly make this a list, not a tensor

File: neuralop/models/uno.py [hunk 6]
```diff
@@ -153,19 +153,25 @@ def __init__(self,
...
+        for k in self.uno_scalings:
```
Comment (JeanKossaifi) [inline neuralop/models/uno.py:165]: Can you add some comment here, it's hard to read and understand the logic.

File: neuralop/models/uno.py [hunk 7]
```diff
@@ -153,19 +153,25 @@ def __init__(self,
...
+            self.output_scaling_factor = np.multiply(self.output_scaling_factor, k)
+        self.output_scaling_factor = self.output_scaling_factor.tolist()
```
Comment (JeanKossaifi) [inline neuralop/models/uno.py:167]: Let's make it a list and not add additional `tolist`.
You can replace the np.multiply (which should work on a list anyway) with math.prod.
Actually we don't need the list, you can directly take the prod I think

File: neuralop/models/spectral_convolution.py [hunk 8]
```diff
@@ -480,8 +480,8 @@ def __init__(self, in_channels, out_channels, n_modes, incremental_n_modes=None,
         
         hm1, hm2 = self.half_n_modes[0], self.half_n_modes[1]
         if self.factorization is None:
-            self.W1 = nn.Parameter(torch.empty(self.in_channels, hm1, hm2, hm1, hm2, dtype = torch.cfloat))
-            self.W2 = nn.Parameter(torch.empty(self.in_channels, hm1, hm2, hm1, hm2, dtype = torch.cfloat))
+            self.W1 = nn.Parameter(torch.empty(hm1, hm2, hm1, hm2, dtype = torch.cfloat))
```
Comment (JeanKossaifi) [inline neuralop/models/spectral_convolution.py:483]: Let's add this in a separate PR

File: neuralop/models/padding.py [hunk 9]
```diff
@@ -54,36 +54,31 @@ def pad(self, x):
...
-
-#             for scale_factor in self.output_scale_factor:
-#                 if isinstance(scale_factor, (float, int)):
-#                     scale_factor = [scale_factor]*len(resolution)
             output_pad = [int(round(i*j)) for (i,j) in zip(self.output_scale_factor,output_pad)]
 
             if self.padding_mode == 'symmetric':
                 # Pad both sides
-                unpad_indices = (Ellipsis, ) + tuple([slice(p, -p, None) for p in output_pad ])
+                unpad_indices = (Ellipsis, ) + tuple([slice(p, -p, None) for p in output_pad[::-1] ])
```
Comment (JeanKossaifi) [inline neuralop/models/padding.py:64]: Why did you reverse here?

File: neuralop/models/resample.py [hunk 10]
```diff
@@ -47,10 +47,10 @@ def resample(x, res_scale, axis, output_shape = None):
...
         new_size = output_shape
 
-    if len(axis) == 1:
-        return F.interpolate(x, size = new_size[0], mode = 'linear', align_corners = True)
-    if len(axis) == 2:
-        return F.interpolate(x, size = new_size, mode = 'bicubic', align_corners = True, antialias = True)
+#     if len(axis) == 1:
+#         return F.interpolate(x, size = new_size[0], mode = 'linear', align_corners = True)
+#     if len(axis) == 2:
+#         return F.interpolate(x, size = new_size, mode = 'bicubic', align_corners = True, antialias = True)
```
Comment (JeanKossaifi) [inline neuralop/models/resample.py:42]: Why did you comment these?

Comment (JeanKossaifi) [reply neuralop/models/resample.py:42]: Can we update to still use them, they are faster

**Review by JeanKossaifi — COMMENTED** [review]
Thanks @ashiq24, left some comments

File: neuralop/models/uno.py (hunk 2 above)
Comment (ashiq24) [reply neuralop/models/uno.py:202]: The Fno blocks takes scaling_factors which are 2d list
Scaling factor for each layer and scaling factor along each dims. 
It can also take a scaler and also 1D list with scaling factor along each layer.

File: neuralop/models/uno.py (hunk 3 above)
Comment (ashiq24) [reply neuralop/models/uno.py:193]: removed it.

File: neuralop/models/uno.py (hunk 4 above)
Comment (ashiq24) [reply neuralop/models/uno.py:232]: done

File: neuralop/models/uno.py (hunk 7 above)
Comment (ashiq24) [reply neuralop/models/uno.py:165]: math.prod multiply numbers in a list. but I need to multiply 2 lists. Each list specifies the scaling factors along each of the dims.

File: neuralop/models/padding.py (hunk 9 above)
Comment (ashiq24) [reply neuralop/models/padding.py:64]: The first element of the output_pad pads around the last dimension. So, the first element should in the last slice.

File: neuralop/models/resample.py (hunk 10 above)
Comment (ashiq24) [reply neuralop/models/resample.py:42]: I can update this. But then the neural operator can only be used with PyTorch version later than 1.9.1. Because, " antialias" is not available in the previous distribution of PyTorch.

File: neuralop/models/padding.py (hunk 9 above)
Comment (JeanKossaifi) [reply neuralop/models/padding.py:64]: Why is that the case? Shouldn't it be in the same order?

File: neuralop/models/uno.py (hunk 7 above)
Comment (JeanKossaifi) [reply neuralop/models/uno.py:166]: Hmm this code is becoming very confusing - can we try to think and simplify this scaling API?

File: neuralop/models/padding.py (hunk 9 above)
Comment (ashiq24) [reply neuralop/models/padding.py:64]: As per my understanding from the documentation, it is (left, right, top, botton, front, back) for 3D tensor... Left/right padding pads along the dim = -1. Top/bottom pads along the dim = -2... and so on.

File: neuralop/models/uno.py (hunk 7 above)
Comment (ashiq24) [reply neuralop/models/uno.py:166]: To do the scaling we need two things ---
one -- scaling factor for each layer.
two -- scaling factor along each of the dimension.
We need a two array in general.
But in case, we scale each of the dimensions with same scaling factor, then it can be a 1d array, where each element is the scaling factor for each layer.

File: neuralop/models/padding.py (hunk 9 above)
Comment (JeanKossaifi) [reply neuralop/models/padding.py:64]: Great point @ashiq24 - can we make a quick comment and point to the doc for this?

File: neuralop/models/uno.py (hunk 7 above)
Comment (JeanKossaifi) [reply neuralop/models/uno.py:166]: Alright, let's process it to always be a nested list, at the highest level and we can use zip to do the element-wise multiplication. 

Do we **need** to have both output_scaling_factor and uno_scaling?

File: neuralop/models/padding.py (hunk 9 above)
Comment (ashiq24) [reply neuralop/models/padding.py:64]: You meant in the code as a doc string?

Comment (JeanKossaifi) [reply neuralop/models/padding.py:64]: Yes, in the code just with `# comment` :)

File: neuralop/models/uno.py (hunk 1 above)
Comment (ashiq24) [reply neuralop/models/uno.py]: done

File: .gitignore [hunk 11]
```diff
@@ -79,3 +79,5 @@ target/
 
 #Ipython Notebook
 .ipynb_checkpoints
+
+setup_git.sh
```
Comment (JeanKossaifi) [inline .gitignore:83]: I don't think this needs to be in the main repo

**Comment by JeanKossaifi** [comment]
Thanks @ashiq24 looks good overall but looks like the latest PR merged created a conflict, can we resolve?

**Comment by ashiq24** [comment]
I think it now suitable for merging.

File: neuralop/models/padding.py (hunk 9 above)
Comment (JeanKossaifi) [reply neuralop/models/padding.py:65]: @ashiq24  can you add a short comment here?

File: .gitignore (hunk 11 above)
Comment (JeanKossaifi) [reply .gitignore:83]: Let's remove this one.

File: neuralop/models/resample.py (hunk 10 above)
Comment (JeanKossaifi) [reply neuralop/models/resample.py:42]: Makes sense, let's open an issue to address this in a separate PR.

File: neuralop/models/uno.py (hunk 5 above)
Comment (JeanKossaifi) [reply neuralop/models/uno.py:163]: @ashiq24 -- a list would be better here

File: neuralop/models/uno.py (hunk 6 above)
Comment (JeanKossaifi) [reply neuralop/models/uno.py:164]: Let's add some comments before merging.

File: neuralop/models/uno.py (hunk 7 above)
Comment (JeanKossaifi) [reply neuralop/models/uno.py:166]: ?

**Review by JeanKossaifi — COMMENTED** [review]
Looks good overall, let's just address the comments to make the code a little clearer before merging.

**Comment by JeanKossaifi** [comment]
Thanks @ashiq24, it looks good - just one last thing, can you have a look at [pep8](https://peps.python.org/pep-0008/)? 
In particular, can you remove the spaces around the ` = `, at least in the function calls/signatures?
Good to merge aside from that.

**Comment by JeanKossaifi** [comment]
Thanks @ashiq24, merging.
